File: dashboard/content-review.js
/* Pass-two candidate review uses saved model evidence and explicit admin choices. */
const list = $('#candidate-list');
let contentReady = false;

function showReviewLoading(running, state) {
  /* Keep the overlay in sync after a click, poll, or page reload. */
  $('#content-loading').hidden = !running;
  if (running && state) {
    $('#content-loading-progress').textContent = reviewProgress(state);
  }
}

function reviewProgress(state) {
  /* Show the active stage rather than the later pair count during extraction. */
  if (state.units_read < state.units_total) {
    return `Extracting: ${state.documents_read}/${state.documents} documents read · ${state.units_read}/${state.units_total} pages, images or sheets. Completed reads are saved.`;
  }
  return `Checking possible duplicates: ${state.pairs_screened}/${state.pairs_total} pairs screened in batches. Completed checks are saved.`;
}

/* A changed decision needs a fresh completion check. */
function resetContentCheck() {
  contentReady = false;
  $('#content-next').hidden = true;
  $('#content-check-note').textContent = '';
}

function paragraphs(items) {
  const box = node('ul', 'content-evidence');
  for (const item of items || []) box.append(node('li', '', item));
  if (!box.children.length) box.append(node('li', '', 'None recorded.'));
  return box;
}

function documentCard(document) {
  const card = node('section', 'content-document');
  card.append(node('h3', '', document.name), node('small', '', document.path));
  const link = node('a', 'button secondary', 'Open original');
  link.href = `/api/content-file?id=${encodeURIComponent(document.id)}`;
  link.target = '_blank'; link.rel = 'noopener'; card.append(link);
  if (/\.(docx|xlsx)$/i.test(document.name)) {
    const details = node('details', 'visual-document');
    details.append(node('summary', '', 'View document'));
    const preview = node('div', 'content-office-preview');
    const controls = node('div', 'page-controls'), previous = node('button', '', 'Previous');
    const count = node('span'), next = node('button', '', 'Next');
    controls.append(previous, count, next); details.append(preview, controls); card.append(details);
    let page = 0, info;
    const load = async () => {
      try {
        if (!info) info = await api(`/api/document?content_id=${encodeURIComponent(document.id)}`);
        renderOfficePreview(preview, await api(`/api/office-view?content_id=${encodeURIComponent(document.id)}&page=${page}`));
        count.textContent = info.labels?.[page] || `Page ${page + 1} of ${info.pages}`;
        previous.disabled = page === 0; next.disabled = page >= info.pages - 1;
      } catch (error) {preview.replaceChildren(node('p', '', error.message)); previous.disabled = next.disabled = true;}
    };
    details.ontoggle = () => {if (details.open) load();};
    previous.onclick = () => {page--; load();}; next.onclick = () => {page++; load();};
    return card;
  }
  for (let n = 0; n < document.units.length; n++) {
    const unit = document.units[n];
    const details = node('details');
    details.append(node('summary', '', unit.label));
    if (unit.image) {
      const picture = node('img', 'content-preview');
      picture.src = `/api/content-image?id=${encodeURIComponent(document.id)}&unit=${n}`;
      picture.alt = `${document.name}, ${unit.label}`;
      details.append(picture);
    }
    if (unit.text) details.append(node('pre', '', unit.text));
    card.append(details);
  }
  return card;
}

function candidateCard(pair) {
  const card = node('article', 'settings-card content-pair');
  const result = pair.comparison;
  card.append(node('h3', '', `${pair.left.name} / ${pair.right.name}`),
              node('p', '', result ? `${result.classification} · ${result.confidence} confidence` : 'Full comparison pending'));
  card.append(node('p', 'content-reason', `Candidate reason: ${pair.reason}`));
  const documents = node('div', 'content-documents');
  documents.append(documentCard(pair.left), documentCard(pair.right)); card.append(documents);
  if (result) {
    for (const [title, field] of [['Matching evidence', 'evidence'], ['Differences', 'differences'], ['Limitations', 'limitations']]) {
      card.append(node('h4', '', title), paragraphs(result[field]));
    }
    const decision = pair.decision;
    const reason = node('textarea'); reason.placeholder = 'Reason for this decision or undo'; reason.setAttribute('aria-label', 'Decision reason');
    const decisionBar = node('div', 'content-decision-bar');
    decisionBar.append(node('p', 'content-decision', decision ?
      `Admin decision: ${decision.verdict} by ${decision.reviewer}. ${decision.reason}` : 'Admin decision pending.'));
    if (decision) {
      const undo = node('button', 'button secondary', 'Undo decision');
      undo.type = 'button';
      undo.onclick = async () => {
        try {
          await api('/api/content/undo', {pair: pair.pair, reviewer: $('#admin-name').value.trim(),
            reason: reason.value.trim() || 'Uncertain; returned to pending review'});
          resetContentCheck(); toast('Decision undone; candidate is pending again.'); await refresh();
        } catch (error) { toast(error.message); }
      };
      decisionBar.append(undo);
    }
    card.append(decisionBar);
    card.append(reason);
    const actions = node('div', 'content-actions');
    for (const [verdict, label] of [['keep_both', 'Keep both'], ['keep_left', 'Keep left'], ['keep_right', 'Keep right']]) {
      const button = node('button', verdict === 'keep_both' ? 'button secondary' : 'button dark', label);
      button.type = 'button'; button.disabled = verdict !== 'keep_both' && result.classification !== 'same_document';
      button.onclick = async () => {
        try {
          await api('/api/content/decide', {pair: pair.pair, verdict, reviewer: $('#admin-name').value.trim(), reason: reason.value.trim()});
          resetContentCheck(); toast('Admin decision saved.'); await refresh();
        } catch (error) { toast(error.message); }
      };
      actions.append(button);
    }
    card.append(actions);
    if (result.classification === 'same_document') card.append(node('small', '', 'Keep left/right records the approved survivor. Admin cleanup remains separate.'));
  }
  return card;
}

async function refresh() {
  const state = await api('/api/content-review');
  const message = $('#content-message');
  if (!state.exact_ready) {
    showReviewLoading(false);
    resetContentCheck();
    message.textContent = `Finish exact duplicate review first. ${state.exact_problems.length} issue(s) remain.`;
    $('#content-controls').hidden = true; $('#candidate-section').hidden = true; return;
  }
  message.textContent = state.run_error || (state.prepared ? 'Pass one is complete. Review every candidate below.' : 'Pass one is complete. Prepare the remaining documents to begin pass two.');
  $('#content-controls').hidden = false;
  $('#prepare-content').hidden = state.prepared;
  $('#run-content').hidden = !state.prepared || state.running || contentReady;
  $('#check-content').hidden = !state.prepared || state.running;
  $('#content-running').hidden = !state.running;
  showReviewLoading(state.running, state);
  $('#content-progress').textContent = state.prepared ?
    reviewProgress(state) : 'No content review prepared yet.';
  $('#candidate-section').hidden = !state.prepared;
  $('#candidate-count').textContent = `(${state.pairs.length})`;
  list.replaceChildren(...state.pairs.map(candidateCard));
  if (state.prepared && !state.pairs.length) list.append(node('p', '', 'No candidate pairs yet. Run a review batch to extract and screen the documents.'));
}

async function action(path) {
  if (path === '/api/content/run') showReviewLoading(true);
  try { await api(path, {}); resetContentCheck(); await refresh(); }
  catch (error) { showReviewLoading(false); toast(error.message); }
}

$('#prepare-content').onclick = () => action('/api/content/prepare');
$('#run-content').onclick = () => action('/api/content/run');
$('#check-content').onclick = async () => {
  const button = $('#check-content'); button.disabled = true;
  try {
    const status = await api('/api/workflow-checks');
    const ready = status.steps[2].checked;
    contentReady = ready;
    $('#content-next').hidden = !ready;
    $('#run-content').hidden = ready;
    $('#content-check-note').textContent = ready ? 'Content review passed its checks.' :
      'Content review is still pending. Complete the remaining candidate decisions and run any unfinished batch.';
  } catch (error) { $('#content-check-note').textContent = error.message; }
  finally { button.disabled = false; }
};
api('/api/session').then(data => { token = data.token; return refresh(); }).catch(error => toast(error.message));
setInterval(() => { if (token) refresh().catch(error => toast(error.message)); }, 5000);

File: dashboard/content_review.py
"""Expose pass-two evidence and admin decisions to the local dashboard."""

import threading
from pathlib import Path

from reconciliation.codex_reviewer import BudgetReached, CodexReviewer, ReviewCancelled
from reconciliation.duplicate_workflow import check, fingerprint
from reconciliation.review_settings import load_config, stage_settings
from reconciliation.vision_workflow import active_config, current_inventory, decide as save_decision, load, prepare as prepare_review, run, undo_decision


def work_path(review):
    """Keep each project's content review beside its manifest."""
    return review.manifest_path.parent / "review"


def exact_problems(review):
    """Return pass-one blockers before any content-review action."""
    return check(review.root, review.manifest, review.manifest_path)


def execution_status(review):
    """Report stopped only after process verification and worker finalization."""
    worker = getattr(review, "content_thread", None)
    worker_running = bool(worker and worker.is_alive())
    active = getattr(getattr(review, "content_engine", None), "active_count", 0)
    cancelled = getattr(review, "content_cancel", None)
    requested = bool(cancelled and cancelled.is_set())
    error = getattr(review, "content_error", "")
    running = worker_running or bool(active)
    status = "running" if running else "idle"
    if requested:
        status = "stopping" if worker_running else "stop_failed" if active or error else "stopped"
    if active and not worker_running:
        status = "stop_failed"
        error = error or "Process shutdown could not be verified. Another review cannot start."
    if status == "stopped":
        error = "Review stopped. Completed results are saved; run again to resume."
    return {"running": running, "active_processes": active, "stop_requested": requested,
            "execution_status": status, "run_error": error}


def snapshot(review):
    """Read current pass-two candidates without accepting model output as approval."""
    problems = exact_problems(review)
    work = work_path(review)
    result = {"exact_ready": not problems, "exact_problems": problems[:30],
              "prepared": (work / "index.json").is_file(),
              **execution_status(review), "pairs": []}
    if problems or not result["prepared"]:
        return result
    index, state = load(work)
    if Path(index["manifest"]).resolve() != review.manifest_path.resolve():
        raise ValueError("Prepared review belongs to a different manifest")
    documents = index["documents"]
    for pair, screen in sorted(state["screens"].items()):
        if not screen["candidate"]:
            continue
        left, right = pair.split(":")
        comparison = state["pairs"].get(pair)
        result["pairs"].append({"pair": pair, "left": document_summary(documents[left]),
                                "right": document_summary(documents[right]),
                                "reason": screen["reason"], "comparison": comparison,
                                "decision": state["decisions"].get(pair)})
    result["documents"] = len(documents)
    result["documents_read"] = sum(bool(doc["units"]) and not doc["error"] and
                                   all(f"{digest}:{n}" in state["units"] for n in range(len(doc["units"])))
                                   for digest, doc in documents.items() if doc.get("accepted", True))
    result["units_read"] = len(state["units"])
    result["units_total"] = sum(len(doc["units"]) for doc in documents.values() if doc.get("accepted", True))
    result["pairs_screened"] = len(state["screens"])
    result["comparisons_done"] = len(state["pairs"])
    result["comparisons_total"] = sum(bool(screen["candidate"]) for screen in state["screens"].values())
    eligible = sum(doc.get("accepted", True) for doc in documents.values())
    result["pairs_total"] = eligible * (eligible - 1) // 2
    return result


def document_summary(document):
    """Provide a source link and page labels for one reviewed document."""
    return {"id": document["id"], "name": Path(document["paths"][0]).name,
            "path": document["paths"][0],
            "units": [{"label": unit["label"], "text": unit["text"], "image": bool(unit["image"])}
                      for unit in document["units"]]}


def prepare(review):
    """Prepare local evidence only after exact-copy cleanup is complete."""
    if exact_problems(review):
        raise ValueError("Finish exact duplicate review first")
    work = work_path(review)
    if (work / "index.json").exists():
        raise ValueError("Content review is already prepared")
    prepare_review(review.manifest_path, work, review.config_path)
    return snapshot(review)


def start(review):
    """Run a bounded model batch in the background so the page stays responsive."""
    if exact_problems(review):
        raise ValueError("Finish exact duplicate review first")
    if execution_status(review)["running"]:
        raise ValueError("Content review is already running")
    work = work_path(review)
    if not (work / "index.json").is_file():
        raise ValueError("Prepare content review first")
    index, _ = load(work)
    active_config(index)
    review.content_error = ""
    review.content_cancel = threading.Event()
    review.content_engine = None

    def worker():
        """Resume saved extraction, screening, and comparison work."""
        try:
            index, state = load(work)
            config = load_config(review.config_path)
            engine = CodexReviewer(work, model=config["model"] or None,
                                   max_calls=config["max_calls"], reasoning=config["reasoning"],
                                   cancel_event=review.content_cancel)
            review.content_engine = engine
            engine.stage_choices = stage_settings(config)
            run(work, index, state, engine)
        except BudgetReached:
            pass
        except ReviewCancelled:
            pass
        except Exception as error:
            review.content_error = str(error)

    review.content_thread = threading.Thread(target=worker, daemon=True)
    review.content_thread.start()
    return snapshot(review)


def stop(review):
    """Cancel active model calls and keep all completed review checkpoints."""
    if not execution_status(review)["running"]:
        raise ValueError("No content review is running")
    review.content_cancel.set()
    engine = getattr(review, "content_engine", None)
    if engine is not None:
        engine.cancel()
    return snapshot(review)


def decide(review, pair, verdict, reviewer, reason):
    """Record an explicit admin verdict for one completed candidate comparison."""
    if exact_problems(review):
        raise ValueError("Finish exact duplicate review first")
    if execution_status(review)["running"]:
        raise ValueError("Wait for the current content-review batch to finish")
    work = work_path(review)
    index, state = load(work)
    if Path(index["manifest"]).resolve() != review.manifest_path.resolve():
        raise ValueError("Prepared review belongs to a different manifest")
    current_inventory(index, state)
    save_decision(work, index, state, pair, verdict, reviewer, reason)
    from dashboard import development
    development.capture(review)
    return snapshot(review)


def undo(review, pair, reviewer, reason):
    """Clear one admin verdict and retain an audit entry for the reversal."""
    if exact_problems(review):
        raise ValueError("Finish exact duplicate review first")
    if execution_status(review)["running"]:
        raise ValueError("Wait for the current content-review batch to finish")
    work = work_path(review)
    index, state = load(work)
    if Path(index["manifest"]).resolve() != review.manifest_path.resolve():
        raise ValueError("Prepared review belongs to a different manifest")
    current_inventory(index, state)
    undo_decision(work, index, state, pair, reviewer, reason)
    from dashboard import development
    development.capture(review)
    return snapshot(review)


def source(review, digest):
    """Serve only unchanged documents named in the prepared review."""
    index, _ = load(work_path(review))
    path = Path(index["documents"][digest]["paths"][0])
    if fingerprint(path) != digest:
        raise ValueError("Source changed; refresh the review")
    return path


def image(review, digest, number):
    """Serve only a verified prepared page image."""
    index, _ = load(work_path(review))
    if number < 0:
        raise IndexError("Invalid unit number")
    unit = index["documents"][digest]["units"][number]
    if not unit["image"]:
        raise FileNotFoundError("No image for this unit")
    path = Path(unit["image"])
    if not path.resolve().is_relative_to((work_path(review) / "assets").resolve()):
        raise ValueError("Prepared image escaped review storage")
    return path
